Add Query Users API

jfreden opened the pull request:
This PR adds a new Query Users API:
- `POST /_security/_query/user` 
- `GET /_security/_query/user` 

It follows a similar convention to the [Query API key API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-query-api-key.html). 

Sample request 
```json
POST /_security/_query/user
{
  "query": { 
    "bool": {
      "must": [
        {
          "wildcard": {
            "username": "org-user*"
          }
        }
      ],
      "filter": [
        {
          "prefix": {
            "roles": {
              "value": "super"
            }
          }
        }
      ],
      "must_not": [
        {
          "term": {
            "enabled": "false"
          }
        }
      ]
    }
  },
  "from": 0,
  "size": 1000,
  "sort": [ 
    "username"
  ],
  "search_after": [
    "org-user-2048"
  ]
}
```

The query supports a subset of query types, including `match_all`, `bool`, `term`, `terms`, `prefix`, `wildcard`, and `exists`.

The supported fields are:
```
username
roles
enabled
full_name
email
```

The query also supports pagination-related fields, analogous to the generic [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html).

The response format is similar to that of the [Get Users API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-user.html) with one key difference: the retrieved users are returned as a list of payloads, instead of as a nested object keyed by username. This is more consistent with the format for a paginated response. 
```json
{
  "total": 3,
  "count": 3,
  "users": [
    { 
    "username": "org-user-2049",
    "roles": [
      "superuser"
    ],
    "full_name": "Jack Nicholson",
    "email": "jacknich@example.com",
    "metadata": { "intelligence" : 7 },
    "enabled": true
  }
  ]
}
```

**Excluded**
- Documentation will be added in https://github.com/elastic/elasticsearch/pull/104255
- `with_profile_id` will be added in a follow up PR ES-7599

jfreden committed:
Add query users API

jfreden committed:
fixup! tests

jfreden committed:
Update docs/changelog/104033.yaml

elasticsearchmachine commented:
Hi @jfreden, I've created a changelog YAML for you.

jfreden committed:
fixup! Minor fixes

jfreden commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/PrivilegeTests.java:
```diff
@@ -339,16 +341,11 @@ public void testManageUserProfilePrivilege() {
             "cluster:admin/xpack/security/role/get",
             "cluster:admin/xpack/security/role/delete"
         );
-        verifyClusterActionDenied(
```
This was duplicated so did a drive-by remove.

jfreden commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexFieldNameTranslator.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public class SecurityIndexFieldNameTranslator {
```
I added this class to support prefix-field name matching, since that's used in the query apikey api. Now I only use this for the query user api, where we only have exact field name mappings and they all happen to match exactly what's in the security index, so translations are not required. For future-proofing it could be good to keep this, but I also understand that it looks overly complex. 

To justify this I'm going to use this for the api key api too in a separate PR.

Let me know your thoughts.

n1v0lg replied:
@javanna I don't think @jfreden had immediate plans to work on this, so a PR is more than welcome! @jfreden is on PTO for the next two weeks

jfreden commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java:
```diff
@@ -0,0 +1,101 @@
+            QueryBuilder translaterdQueryBuilder = translateToUserQueryBuilder(queryBuilder);
+            userQueryBuilder.must(translaterdQueryBuilder);
+        }
+        userQueryBuilder.filter(QueryBuilders.termQuery("type", "user"));
```
For apikey this is called `doc_type`, just wanted to confirm that `type` for user is correct and expected here?

jfreden replied:
Thank you for doing that! Looks like this filter is acceptable based on the conversation and your comment here.

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

jfreden committed:
fixup! Remove metadata as valid query field

jfreden committed:
fixup! Also remove metadata from tests

jfreden committed:
fixup! lint

n1v0lg reviewed (commented):
This is looking great! I've focused on the prod so far. I'll make a thorough pass over the tests tomorrow 👀 really nice work 💯

jfreden commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/QueryUserAction.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * ActionType for retrieving Users
+ */
+public final class QueryUserAction extends ActionType<QueryUserResponse> {
```
Nice! Didn't know about that one, will add it there instead.

n1v0lg replied:
Thanks!

jfreden committed:
fixup! Code review comments

jfreden commented on x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:
```diff
@@ -0,0 +1,376 @@
+        final String invalidFieldName = randomFrom("doc_type", "invalid", "password");
+        assertQueryError(400, "{\"sort\":[\"" + invalidFieldName + "\"]}");
+    }
+
```
Ah yes, that's a good thing to test. Will add a test for that.

jfreden committed:
fixup! Code review - add reserved user query

n1v0lg reviewed (commented):
Exciting stuff! A few test suggestions, and one more thing we gotta iron out:

Turns out that sorting on `full_name` and `email` does not work since they are `text` fields (see [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param)) -- I totally missed this when putting together the design doc, sorry! I think we should detect this and return a "friendly" error message. Note that _searching_ on them still works, it's just sorting that doesn't.

jfreden commented:
Thanks for the review! 

> Turns out that sorting on full_name and email does not work since they are text fields (see [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param)) -- I totally missed this when putting together the design doc, sorry! I think we should detect this and return a "friendly" error message. Note that searching on them still works, it's just sorting that doesn't.

Yes, I noticed that too when testing. My thinking was that since the error handling is the same as for regular queries, it should be fine. Thinking about it some more though, the "regular error handling" is for indices that are user created, meaning the user setup the text mapping of the field, in this case we did so it's a little different. I'll add some better error handling for that.

jfreden commented on x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:
```diff
@@ -0,0 +1,403 @@
+
+    private User createRandomUser() throws IOException {
+        return createUser(
+            randomValueOtherThanMany(reservedUsers::contains, () -> randomAlphaOfLengthBetween(3, 8)),
```
Good catch! Will fix.

jfreden committed:
fixup! Code review comments

jfreden committed:
fixup! Test

jfreden committed:
fixup! test

jfreden committed:
fixup! Add more tests

jfreden committed:
fixup! lint

jfreden committed:
Add check for sorting on email and full_name

jfreden committed:
fixup! Add tests

jfreden committed:
fixup! Forbidden API fixes

jfreden commented:
PR should be ready for another round of review @n1v0lg 
- Added a check and some tests for detecting sorting on `full_name` and `email`.
- Addressed your other great suggestions.

Thank you!

jfreden committed:
fixup! Test bug

n1v0lg reviewed (approved):
LGTM 🚀 🚀 🚀 

Only smaller comments, no need for re-review. Also, lets add a `test-serverless-update` label before merging. I don't expect this to break anything in serverless but easier to validate.

jfreden commented on x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryUserIT.java:
```diff
@@ -0,0 +1,490 @@
+            true
+        );
+        String enabledTerm = "\"enabled\":true";
+        String fullNameTerm = "\"full_name\":\"Batman\"";
```
Yes, good catch!

jfreden committed:
fixup! Code review comments
